airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_streams.py

```diff
@@ -0,0 +1,186 @@
+
+import pendulum
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.utils.transform import TransformConfig, TypeTransformer
+from cached_property import cached_property
+from facebook_business.adobjects.abstractobject import AbstractObject
+from facebook_business.api import FacebookAdsApiBatch, FacebookRequest, FacebookResponse
+
+from .common import deep_merge
+
+if TYPE_CHECKING:
+    from source_facebook_marketing.api import API
+
+logger = logging.getLogger("airbyte")
+
+
+class FBMarketingStream(Stream, ABC):
+    """Base stream class"""
+
+    primary_key = "id"
+    transformer: TypeTransformer = TypeTransformer(TransformConfig.DefaultSchemaNormalization)
+
+    page_size = 100
+    use_batch = False
+
+    enable_deleted = False
+    entity_prefix = None
+
+    MAX_BATCH_SIZE = 50
+
+    def __init__(self, api: "API", include_deleted: bool = False, **kwargs):
+        super().__init__(**kwargs)
+        self._api = api
+        self._include_deleted = include_deleted if self.enable_deleted else False
+
+    @cached_property
+    def fields(self) -> List[str]:
+        """List of fields that we want to query, for now just all properties from stream's schema"""
+        return list(self.get_json_schema().get("properties", {}).keys())
+
+    def _execute_batch(self, batch):
+        """Execute batch, retry in case of failures"""
+        while batch:
+            batch = batch.execute()
+            if batch:
+                logger.info("Retry failed requests in batch")
+
+    def execute_in_batch(self, pending_requests: Iterable[FacebookRequest]) -> Iterable[MutableMapping[str, Any]]:
+        """Execute list of requests in batches"""
+        records = []
+
+        def success(response: FacebookResponse):
+            records.append(response.json())
+
+        def failure(response: FacebookResponse):
+            # FIXME: stop sync or retry
+            logger.warning(f"Request failed with response: {response.body()}")
+
+        api_batch: FacebookAdsApiBatch = self._api.api.new_batch()
+        for request in pending_requests:
+            api_batch.add_request(request, success=success, failure=failure)
+            if len(api_batch) == self.MAX_BATCH_SIZE:
+                self._execute_batch(api_batch)
+                yield from records
+                api_batch: FacebookAdsApiBatch = self._api.api.new_batch()
+                records = []
+
+        self._execute_batch(api_batch)
+        yield from records
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        """Main read method used by CDK"""
+        records_iter = self._read_records(params=self.request_params(stream_state=stream_state))
```

Comment:
the default impl of `_read_records` returns an empty list. Doesn't this mean that this method always returns nothing unless `_read_records` is overridden? Doesn't that make it a required method? I'm confused by the interaction between `_read_records` and this `read_records` implementation

Reply:
@sherifnada 
renamed `_read_records` to `list_objects` and made it `abstractmethod` with noop implementation for Insights.
I hope it makes less weird now.